File: 0039-combination-sum/0039-combination-sum.java
class Solution {
    private void find(int idx, int[] arr, int tar, List<List<Integer>> ans, List<Integer> ds) {
        if (idx == arr.length) {
            if (tar == 0)
                ans.add(new ArrayList<>(ds));
            return;
        }
        if (arr[idx] <= tar) {
            ds.add(arr[idx]);
            find(idx, arr, tar - arr[idx], ans, ds);
            ds.remove(ds.size() - 1);
        }
        find(idx + 1, arr, tar, ans, ds);
    }

    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        List<List<Integer>> ans = new ArrayList<>();
        find(0, candidates, target, ans, new ArrayList<>());
        return ans;
    }
}

File: 0039-combination-sum/0039-combination-sum.go
func find(idx int, arr []int, tar int, ans *[][]int, ds *[]int) {
    if tar == 0 {
        combination := make([]int, len(*ds))
        copy(combination, *ds)
        *ans = append(*ans, combination)
        return
    }
    
    if idx == len(arr) {
        return
    }

    if arr[idx] <= tar {
        *ds = append(*ds, arr[idx])
        find(idx, arr, tar - arr[idx], ans, ds)
        *ds = (*ds) [ : len(*ds) - 1]
    }

    find(idx + 1, arr, tar, ans, ds)
}

func combinationSum(candidates []int, target int) [][]int {
    ans := make([][]int, 0)
    ds := make([]int, 0)

    find(0, candidates, target, &ans, &ds)
    return ans
}